pkg/receive/handler.go

```diff
@@ -533,28 +533,45 @@ func (h *Handler) replicate(ctx context.Context, tenant string, wreq *prompb.Wri
 	}
 	h.mtx.RUnlock()
 
-	var err error
-	ctx, cancel := context.WithTimeout(ctx, h.options.ForwardTimeout)
-	defer func() {
-		// If there is no error, let forward requests optimistically run until timeout.
+	var (
+		quorum = h.writeQuorum()
+		// NOTICE: The cancel function is not used on all paths intentionally,
+		// to let forward requests to optimistically run until timeout,
+		// which is why the nolint comment is required and intentional here.
+		fctx, cancel = context.WithTimeout(context.Background(), h.options.ForwardTimeout) //nolint:govet
+		errChan      = make(chan error)
+	)
+	go func() {
+		defer close(errChan)
+		errChan <- h.fanoutForward(fctx, tenant, replicas, wreqs, quorum)
+	}()
+
+	select {
+	case err := <-errChan:
 		if err != nil {
+			// fanoutForward only returns an error if successThreshold (quorum) is not reached.
+			// Cancel all in-flight requests.
 			cancel()
+
+			if countCause(err, isNotReady) >= quorum {
+				return errors.Wrap(tsdb.ErrNotReady, "replicate: quorum not reached")
+			}
+			if countCause(err, isConflict) >= quorum {
+				return errors.Wrap(conflictErr, "replicate: quorum not reached")
+			}
+
+			return errors.Wrap(err, "unexpected error")
 		}
-	}()
 
-	quorum := h.writeQuorum()
-	err = h.fanoutForward(ctx, tenant, replicas, wreqs, quorum)
-	if countCause(err, isNotReady) >= quorum {
-		return tsdb.ErrNotReady
-	}
-	if countCause(err, isConflict) >= quorum {
-		return errors.Wrap(conflictErr, "did not meet success threshold due to conflict")
+		// If there is no error, quorum reached, let forward requests to optimistically run until timeout.
+		// NOTICE: This return statement may be reached without using the cancel var defined on line 540,
+		// this is the intended behavior since we want to forward requests to optimistically run until timeout,
+		// which is why the nolint comment is required and intentional here.
+		return nil //nolint:govet
```

Comment:
so... why nolint is here if we removed go vet?

Reply:
We didn't remove the `go vet`. We removed redundant `go vet`. `golangci-lint` runs it already, we were running it again as a separate command.